tests/core/data/test_network.py

```diff
@@ -0,0 +1,334 @@
+import numpy as np
+from numpy.testing import assert_almost_equal
+import pytest
+
+from qutip.core import data
+from qutip.core.data import dense
+from qutip_tensornetwork.core.data import Network
+from qutip_tensornetwork.testing import assert_network_close
+from .conftest import random_node, random_complex_network
+import tensornetwork as tn
+
+def test_init_only_in_edge():
+    node = random_node((2,))
+    network = Network([], node[0:], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
+    assert len(network.in_edges) == 1
+    assert network.in_edges[0] is node[0]
+    assert len(network.out_edges) == 0
+
+def test_init_only_out_edge():
+    node = random_node((2,))
+    network = Network(node[0:], [], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
+    assert len(network.out_edges) == 1
+    assert network.out_edges[0] is node[0]
+    assert len(network.in_edges) == 0
+
+def test_init_both_in_and_out_edge():
+    node = random_node((2, 2))
+    network = Network(node[0:1], node[1:], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
+    assert len(network.out_edges) == 1
+    assert network.out_edges[0] is node[0]
+    assert len(network.in_edges) == 1
+    assert network.in_edges[0] is node[1]
+
+def test_init_raise_if_edges_not_unique():
+    """in_edges and out_edges must be unique."""
+    node = random_node((2, 2))
+    with pytest.raises(ValueError):
+       network = Network([node[0]], [node[0]])
+
+def test_non_dangling_in_or_out_raises():
+    """Input edges for in_edges and out_edges must be dangling."""
+    node = random_node((2,))
+    node2 = random_node((2,))
+    node[0] ^ node2[0]
+
+    with pytest.raises(ValueError):
+       network = Network([node[0]], [])
+
+    with pytest.raises(ValueError):
+       network = Network([], [node[0]])
+
+def test_nodes_dont_include_edges_raises():
+    """Included nodes do not have the passed in_edges or out_edges."""
+    node = random_node((2,2))
+    node2 = random_node((2,2))
+    with pytest.raises(ValueError):
+        network = Network(node[0:1], node[1:], nodes=[node2], copy=False)
+
+def test_non_in_out_edges_dangling_raises():
+    """Included nodes can not have dangling edges that are not in_edges or
+    out_edges
+    """
+    node = random_node((2,2,2))
+    with pytest.raises(ValueError):
+        network = Network([node[0]], [node[1]], copy=False)
+
+def test_scalar_nodes():
+    """It is possible to include scalar nodes as long as they are not the only
+    nodes.
```

Comment:
This comment is a bit confusing. It says "as long as they are not the only nodes" but test_scalar_network successfully builds a network with only a scalar node. Perhaps the wording just needs to change a bit?

Reply:
`test_scalar_network` builds the network with an scalar node, `node_scalar`, and a non-scalar node, `node`.

Reply:
At first I was not going to allow the creation of scalar networks but now I do. This test should be removed.